### DISCUSSION TITLE
How to automate clientSecret (APPLE_SECRET) generation?

### DISCUSSION THREAD

**Discussion by misterioss** [nextauthjs/next-auth#9420] [Help]
For now, I need to recreate clientSecret for Apple Auth every six month. Is there a way or documentation to make this process automatic? 
Script for JWT generation:
(https://gist.github.com/balazsorban44/09613175e7b37ec03f676dcefb7be5eb)

**Comment by leo3linbeck** [comment]
I use the following function to regenerate the Apple Secret (which is a JWT) each time the user re-authenticates (which is not that often, given that the auth tokens are cached locally and re-used until they expire):

```
import { SignJWT, importPKCS8 } from 'jose';

const alg = 'ES256';
const kid = `${AUTH_APPLE_KEY_ID}`;
const pk = await importPKCS8(`${AUTH_APPLE_PRIVATE_KEY}`.replace(/\\n/g, '\n'), alg);

const appleSecret = await new SignJWT({})
	.setAudience('https://appleid.apple.com')
	.setIssuer(`${AUTH_APPLE_TEAM_ID}`)
	.setIssuedAt()
	.setExpirationTime('1 year')
	.setSubject(`${AUTH_APPLE_CLIENT_ID}`)
	.setProtectedHeader({ alg, kid })
	.sign(pk);

```
I put this code in the auth.ts file, prior to invoking the Auth function, so it runs once when the module is executed.

The appleSecret value then is passed as the client secret to the Apple provider:

```
Apple({
     clientId: `${AUTH_APPLE_CLIENT_ID}`,
     clientSecret: appleSecret
})

```

You'll need to install jose using npm.

Also, the private key (the one Apple generates for you in the keys section of the Certificates, Identities & Profiles section of the Apple Developer website)  needs to be stored as an environment variable. This can be done by changing all of the newline characters to \n (two plain characters) so it will be on a single line in the .env file. The code above changes the \n back to a newline character, which is needed by the importPKCS8 library function. The other environment variable listed above also come from the Apple site - CLIENT_ID is App ID, TEAM_ID is the Team ID (duh), and KEY_ID is the Key ID of the primary key you generated and use to generate the JWT that is the appleSecret.

BTW: I still can't get the Apple sign in to work for my use case, which requires redirect, which in turn requires state validation to access the email returned by Apple, which fails because Apple returns the state in a POST body and Authjs expects state to be returned as a query parameter. I'm sure it will get fixed at some point, maybe even by me if I can get past our product launch! ;-)

Hope this helps.

Cheers,
L3
